### PR TITLE
Dht storage

### PR DESCRIPTION
**Pull request by seun-ja**
DHT features are now active. Most of the work done is in node/src/service.rs
- I have included the finalised blocks as the value, while the finalised block index as the key.
- With that, finalised blocks are the stored value, and it's the data retrieved on presenting the key
- Also updated here is the inherent data passed into the block for consensus purposes. 
- It now uses the data passed into the DHT for consensus.
   - It can be noticed by testing its replication factor.
- Readability has also been improved.

### CODE REVIEW

File: client/src/lib.rs [hunk 1]
```diff
@@ -4,14 +4,12 @@ use core::fmt;
...
+use frame_system::{ensure_none, ensure_signed, offchain::{
 		AppCrypto, CreateSignedTransaction,
-		SignedPayload, Signer, SigningTypes, SubmitTransaction,
+		SignedPayload, SigningTypes, SubmitTransaction,
 	}};
 use sp_core::{crypto::KeyTypeId};
-use sp_core::offchain::OpaqueMultiaddr;
 use sp_io::offchain_index;
-use sp_api;
 use sp_runtime::{generic,
```
Comment (pratiktri) [inline client/src/lib.rs:13]: General readability improvement suggestions:-

- Club all non-public `const` together
- Documentation comments (`\\\`) on all enum variants
- Remove commented out code
- Run `cargo fmt` to format the code

File: client/src/lib.rs [hunk 2]
```diff
@@ -23,7 +21,7 @@ use sp_runtime::{generic,
 	},
 	RuntimeDebug,
 };
-use sp_runtime::traits::{BlakeTwo256, Block};
+use sp_runtime::traits::{BlakeTwo256};
 use sp_std::{collections::vec_deque::VecDeque, prelude::*, str};
```
Comment (pratiktri) [inline client/src/lib.rs:26]: Please run `cargo clippy` which points out some redundant `clone()` and other issues.

File: node/src/service.rs [hunk 3]
```diff
@@ -34,6 +29,45 @@ type FullClient = sc_service::TFullClient<Block, RuntimeApi, Executor>;
...
+		// let key = format!("{:?}", self.key);
+		//Todo: convert keys above to bytes array. 
+		&INHERENT_IDENTIFIER
+    }
+
+    fn provide_inherent_data(&self, inherent_data: &mut sp_consensus::InherentData) -> Result<(), sp_inherents::Error> {
+        // dht value: block
+		// let mut data = InherentData::new();
+		inherent_data.put_data(INHERENT_IDENTIFIER, &self.value)
+		// todo!()
```
Comment (pratiktri) [inline node/src/service.rs:50]: Remove this line, since we have already provided an implementation

File: node/src/service.rs [hunk 4]
```diff
@@ -34,6 +29,45 @@ type FullClient = sc_service::TFullClient<Block, RuntimeApi, Executor>;
...
+
+    fn error_to_string(&self, error: &[u8]) -> Option<String> {
+        sp_inherents::Error::decode(&mut &error[..]).map(|e| e.into_string()).ok()
+    }
+}
+
+pub fn build_inherent_data_providers(key: String) -> Result<InherentDataProviders, ServiceError> {
+	let providers = InherentDataProviders::new();
+
+	let key_new= key.into_bytes();
```
Comment (pratiktri) [inline node/src/service.rs:61]: Can we avoid creating this variable and use `key.into_bytes()` inside `register_provider`?

File: node/src/command.rs [hunk 5]
```diff
@@ -23,7 +23,7 @@ use node_template_runtime::Block;
 
 impl SubstrateCli for Cli {
 	fn impl_name() -> String {
-		"Substrate Node".into()
+		"S-O Substrate Node".into()
```
Comment (pratiktri) [inline node/src/command.rs:26]: I'm 100% sure where this is being displayed, but can we give it a more descriptive name?

File: client/src/lib.rs (hunk 1 above)
Comment (seun-ja) [reply client/src/lib.rs:13]: can I get clarity on the first point, do you mean I should put all the non-public `const` code together?

File: node/src/service.rs (hunk 4 above)
Comment (seun-ja) [reply node/src/service.rs:61]: spot-on

File: node/src/command.rs (hunk 5 above)
Comment (seun-ja) [reply node/src/command.rs:26]: I was using it to test, but I'll revert to original state

File: client/src/lib.rs (hunk 1 above)
Comment (pratiktri) [reply client/src/lib.rs]: Yes, basically all private `const` together without spaces between them. Something like:-

```rust
/// The keys can be inserted manually via RPC (see `author_insertKey`).
pub const KEY_TYPE: KeyTypeId = KeyTypeId(*b"demo");

// The type to sign and send transactions.
const UNSIGNED_TXS_PRIORITY: u64 = 100;
// TODO: api request
const HTTP_REMOTE_REQUEST: &str = "https://api.pro.coinbase.com/products/ETH-USD/ticker";
const HTTP_HEADER_USER_AGENT: &str = "jaminu71@gmail.com";
const HTTP_ETHEREUM_HOST: &str = "http://127.0.0.1:8080";
const FETCH_TIMEOUT_PERIOD: u64 = 3000; // in milli-seconds
const LOCK_TIMEOUT_EXPIRATION: u64 = FETCH_TIMEOUT_PERIOD + 1000; // in milli-seconds
const LOCK_BLOCK_EXPIRATION: u32 = 3; // in block number
const ONCHAIN_TX_KEY: &[u8] = b"ocw-demo::storage::tx";
```

**Pull request merged**